katdal/visdatav4.py

```diff
@@ -277,24 +279,42 @@ def __init__(self, source, ref_ant='', time_offset=0.0, applycal='',
         self.sensor['Observation/label'] = label
         self.sensor['Observation/compscan_index'] = CategoricalData(list(range(len(label))),
                                                                     label.events)
-        # Use the target sensor of reference antenna to set target for each scan
-        target = self.sensor.get(self.ref_ant + '_target')
-        # Remove initial blank target (typically because antenna starts stopped)
-        if len(target) > 1 and target[0] == 'Nothing, special':
-            target.events, target.indices = target.events[1:], target.indices[1:]
-            target.events[0] = 0
+        # Use the relevant target sensor to set target for each scan
+        ref_target = self.ref_ant or 'cbf'
+        target = self.sensor.get(ref_target + '_target')
         # Move target events onto the nearest scan start
         # ASSUMPTION: Number of targets <= number of scans
         # (i.e. only a single target allowed per scan)
         target.align(scan.events)
+        # Remove initial target if antennas start in mode STOP
+        # (typically left over from previous capture block)
+        for segment, scan_state in scan.segments():
+            # Keep going until first non-STOP scan or a new target is set
+            if scan_state == 'stop' and target[segment.start] is target[0]:
+                continue
+            # Only remove initial target event if we move to a different target
+            if target[segment.start] is not target[0]:
+                target.events = target.events[1:]
+                target.indices = target.indices[1:]
+                target.events[0] = 0
+                target.align(target.events)
+            break
         self.sensor['Observation/target'] = target
         self.sensor['Observation/target_index'] = CategoricalData(target.indices,
                                                                   target.events)
-        # Set up catalogue containing all targets in file, with reference
-        # antenna as default antenna
+        # Set up catalogue containing all targets in file
         self.catalogue.add(target.unique_values)
-        ref_sensor = 'Antennas/%s/antenna' % (self.ref_ant,)
-        self.catalogue.antenna = self.sensor.get(ref_sensor)[0]
+        if self.ref_ant in cam_ants:
+            # If provided, use reference antenna as default antenna
+            cat_ant = [ant for ant in ants if ant.name == self.ref_ant][0]
```

Comment:
The `self.ref_ant in cam_ants` check takes care of that. Line 309 is simply a way to turn an antenna name into an Antenna object without building a dict first.

Reply:
Oh right, I'd missed that above. But I think that check should rather be `if self.ref_ant`: if the user specified a refant, you should try to use it, and report an error if you couldn't, rather than silently using the array centre instead.

Reply:
Agreed, that's a good point.

Reply:
I now check `ref_ant` against `cam_ants` up front (otherwise the first trouble will be a missing `xxx_activity` sensor KeyError...) and build an antenna dict on the fly later.

I'm using `cam_ants` instead of `obs_ants` because for partitioning and catalogue purposes the only real requirement is the presence of the appropriate sensors and an Antenna object.